ohai/gpu_devices.rb

```diff
@@ -0,0 +1,24 @@
+Ohai.plugin(:GPUDevices) do
+  provides 'gpu_devices'
+
+  def gpu_devices_from_lspci
+    shell_out('lspci -vmm').stdout.split("\n\n").select do |pcidev|
+      pcidev =~ /(?:3D|VGA)(?: compatible)? controller/
+    end.map do |gpudev|
+      Mash.new.tap do |properties|
+        gpudev.lines.map do |line|
+          line.chomp!
+          key, val = line.split(":\t")
+          properties[key.downcase] = val
+        end
+      end
+    end
```

Comment:
This is pretty magical ruby, but Ohai makes it somewhat hard to avoid.

The output from lspci's verbose, modern, machine-readable (`-vmm`) is in not-quite-deb822 formatted stanzas, one per device. I split on double newlines to break up the device stanzas and then filter out the ones which match a GPU device regex.

The regex was crafted based on existing observations of various hosts and instance types. This would technically match `3D compatible controller` which is not a class I've seen in the wild but would probably still be a GPU anyway, so I'm not bothered.

The individual key value lines are then split by the separator string `:<tab>` and then placed in a new Mash containing the properties for that gpu device. The `tap` method is a common Ruby idiom which returns the tapped object after manipulating it within the block. The use of this idiom just prevents accidentally returning some other value to the outer `map` block instead of the newly created Mash.